Added option to validate which models are allowed to be associated to a polymorphic belongs_to

natematykiewicz opened the pull request:
### Motivation / Background

When using a polymorphic belongs_to, I usually have a list of allowed classes that I expect this association to use. Oftentimes there's even certain views that need to handle each expected option. Unexpected types in the table would cause problems and potentially pose a security risk.

I've manually added an inclusion validator before, but wished I could just tell the association what the allowed options are.

### Detail

This Pull Request changes `belongs_to`'s `polymorphic` option to allow an array of Strings/Symbols of allowed model names. When provided, an inclusion validator is added to the model to ensure the `_type` field is in the list of allowed class names.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

flavorjones reviewed (changes requested):
I like this feature and the code and tests look good to me. 

Presuming a Core team member will also approve, can you also add something to the "Polymorphic Associations" section in `guides/source/association_basics.md`?

natematykiewicz commented:
> I like this feature and the code and tests look good to me.
> 
> Presuming a Core team member will also approve, can you also add something to the "Polymorphic Associations" section in `guides/source/association_basics.md`?

Done! https://github.com/rails/rails/pull/51738/files#diff-d977574b785ce95fe656c58211c668992bf95214d52d3ffadb409ef8f69ad287

I've not done many references in markdown. I tried to follow other parts of the guide. I *think* I linked the text to https://edgeguides.rubyonrails.org/active_record_validations.html#inclusion correctly. I'm not quite sure of the best way to preview the guides, so I'm sort of flying blind on that.

flavorjones reviewed (approved):
This looks good to me. Needs to be rebased but as soon as you do that, I'll tag it for review by one of the core team.

natematykiewicz commented:
@flavorjones I've rebased the branch

natematykiewicz committed:
Add option to validate which models are allowed to be associated to a polymorphic belongs_to

natematykiewicz commented:
Just fixed another merge conflict on the changelog. The ActiveRecord changelog is a hot file. I guess this is why the Rubocop repo has PRs add a new file to a changelogs folder, and then they generate a CHANGELOG.md at each release.

flavorjones commented:
@natematykiewicz Thanks! I've poked the core team to review.

rafaelfranca commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -121,6 +121,13 @@ def self.define_validations(model, reflection)
       end
 
+      case reflection.options[:polymorphic]
+      when Array, String, Symbol
```
Should we accept String and Symbol? If there is only one possible value people should just pass an array with one element

natematykiewicz replied:
Are you thinking we simply support arrays of strings? I was trying to be flexible, but only supporting an array of strings is certainly easier to document and explain.

natematykiewicz replied:
Or are you saying I just support an array, and then map the values to string still (which would allow an array of strings/symbols)?

flavorjones replied:
I think Rafael is suggesting the latter: just support `Array[String|Symbol]` and not support passing a single string or symbol.

natematykiewicz replied:
I've made the changes, but I'm not sure what to do about the changelog. I see @rafaelfranca changed the milestone to 8.0.0. So I need to rebase this to main's recently wiped changelog instead of 7.2.0's changelog?

flavorjones replied:
@natematykiewicz Yes, please do add it to the new (empty or nearly-empty) CHANGELOG file, this would now land in 8.0.

natematykiewicz replied:
Done @flavorjones!

flavorjones commented:
@natematykiewicz Please check the rubocop failure on the latest push, https://buildkite.com/rails/rails/builds/107357#018f8f2d-ae4b-4d92-8ca4-579e697c5149/1253-1260

natematykiewicz commented:
Sorry about that. I've fixed it. @flavorjones

natematykiewicz commented:
FYI, the buildkite failure is unrelated:

```
Failure:
ActiveRecord::ConnectionAdapters::ReaperTest#test_connection_pool_starts_reaper_in_fork [test/cases/reaper_test.rb:161]:
Expected #<Process::Status: pid 362 SIGABRT (signal 6) (core dumped)> to be success?.
```

natematykiewicz commented:
I just fixed the merge conflict on the changelog

seanpdoyle commented:
This overlaps a little with the [delegated_type](https://edgeapi.rubyonrails.org/classes/ActiveRecord/DelegatedType.html#method-i-delegated_type) declaration that was recently added. That class method accepts a `:types` keyword of `Array[String]`. Under the hood, [it declares a `belongs_to`](https://github.com/rails/rails/blob/26f8edf411bee476049507cab2e59e4dffafca30/activerecord/lib/active_record/delegated_type.rb#L232-L233), then forwards the `:type` option to a another method that defines the attributes and predicates.

Is there an opportunity to make the interfaces more consistent? Is there an opportunity for this PR to also change the `delegated_type` implementation to rely on the new validations?

natematykiewicz commented:
@seanpdoyle I was thinking the exact same thing, but figured maintainers would want 2 PRs. I'd gladly add it to this one, though.

natematykiewicz commented:
@seanpdoyle I was thinking the exact same thing, but figured maintainers would want 2 PRs. I'd gladly add it to this one, though.

seanpdoyle commented:
> but figured maintainers would want 2 PRs. I'd gladly add it to this one, though.

👍! I can't weigh-in on this, and definitely defer to maintainers. I just didn't want `delegated_type` to be overlooked!

natematykiewicz commented:
@flavorjones I rebased to main again

@rafaelfranca removed support for String/Symbol and made it only support an Array of Strings and/or Symbols. Is that what you were looking for?

zzak commented:
This is maybe not a strongly shared opinion, but validations are meant for the user, not developers. What is a user going to do with this error?

Would you go through the trouble of putting a constraint on your database for this?

natematykiewicz commented:
My primary use-case is for a JSON API. I do not control the UI, so I can't make it impossible for someone to try to insert wrong values. I don't want 500 errors in my error monitor because someone provided an invalid item_type.

Maybe a lot of people won't find a need for this, but I have like 5 places in my Rails app that this would immediately help me.

I also really like the idea of being able to see the allowed types in my model, rather than having to turn this into a Postgres enum.

zzak commented:
But can't you already add this validation yourself?

```
validates :item_type, inclusion: { in: %w(Post Comment) }
```

Without the need to add this for your specific use-case.

natematykiewicz commented:
That's not quite the same. If the association is non-optional, a wrong "type" will provide two validation errors on one field, which in my opinion is silly.

Also, I find this to be a nice developer experience improvement for a few lines of code. I've frequently wished I could tell the polymorphic association which types are allowed, just like you tell a `delegated_type` what types are allowed.

https://api.rubyonrails.org/classes/ActiveRecord/DelegatedType.html

natematykiewicz commented:
As @seanpdoyle pointed out (and I was also thinking before opening the PR), it could make a lot of sense to make delegated types pass the `types` option to the polymorphic belongs_to association it creates. That's a 1-line change.

zzak commented:
> make delegated types pass the types option to the polymorphic belongs_to association

This and,

> Is there an opportunity for this PR to also change the delegated_type implementation to rely on the new validations?

Are still open questions, which I guess is the point of this PR.

But to make APIs appear to be consistent not such a strong argument IMO.

natematykiewicz commented:
You're right though. I can and most certainly do manually add a validation. 

But I think that it's a nice developer experience improvement for the polymorphic association to be in charge of that validation, automatically using the appropriate foreign key column and automatically applying an `if` to skip validation if the field is blank (deferring the validation to the `optional` true/false option).

Considering that `optional` is a validation, I don't see why we wouldn't also have the appropriate `inclusion` validation baked in.

rafaelfranca commented:
I'm ok with this change because delegated_types use it, but just to point out, you could do this and have the same behavior:

```
validates :item_type, inclusion: { in: %w(Post Comment) }, allow_blank: true
```

rafaelfranca commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -121,6 +121,12 @@ def self.define_validations(model, reflection)
 
+      if reflection.options[:polymorphic].is_a?(Array)
+        model.validates reflection.foreign_type,
+          inclusion: { in: reflection.options[:polymorphic].map { |klass| klass.to_s.freeze } },
```
Why do we need to `freeze` here? Is it to dedup the string and not have a lot of allocations? If that is the case it should not matter, this only run once per call during boot time.

natematykiewicz replied:
Yeah I was trying to avoid string allocations by freezing them. These strings are retained and kept in the validator for the life of the app, right?

rafaelfranca replied:
Yes, but I don't think that matters

natematykiewicz replied:
Okay. I was trying to get these strings to share with `User.name`, since that's frozen. But I can just remove it. I was trying to think of what changes Jean might to this later, and do them first.

natematykiewicz replied:
I've removed the `.freeze`. I also added a sentence to the `belongs_to` API documentation, and rebased.

rafaelfranca commented:
Can you add a new commit for changing deleted_types to use the same feature in this PR? Doesn't need to be two PRs, just two commits

natematykiewicz commented:
I'll work on adding this to delegated types, too. Thanks for the review!

natematykiewicz committed:
Validate delegated_types polymorphic foreign key

natematykiewicz commented:
@rafaelfranca I just added a commit for delegated_types.

I had to change the test `Entry` model. It had 2 delegated types set up for the same columns, so the validator on `entryable_type` was `[Message, Comment]` and then replaced with `[Post]`, which made existing tests fail. I added 2 more columns and separated the delegated types, so now they have independent validations. It seems like the goal of the Post test is just to ensure that foreign_key and foreign_type work, and the previous PR just reused the existing already-used columns. 

https://github.com/rails/rails/blob/f5355a21ae536a6638d503389a549e00c45cfa29/activerecord/test/models/entry.rb#L4-L9

natematykiewicz commented:
Hmm, buildkite failure is related to my changes, but the tests pass locally. I'll get it fixed.

natematykiewicz commented:
I pushed a fix and buildkite passes now

natematykiewicz commented:
@rafaelfranca I saw you rebased this. Do you need anything from me?

dhh commented:
Hmm. Where is this coming from? It smells like type checking to me. Are you using this in some current application? What kinds of bugs has it actively prevented? I'm not convinced this is a good idea.

natematykiewicz commented:
Our app has quite a few polymorphic belongs_tos. Certain associations only allow certain classes to be associated or things will blow up elsewhere, so I need to validate which models are associated somehow. A validation seems like the Rails way to do that. Doing it on the association seems like the best spot, similar to how `delegated_type`s declare what models can be delegated.

Additionally, one problem I've experienced is that ActiveRecord automatically `const_get`s whatever string was handed into the `_type` polymorphic belongs_to field. This has resulted in 500 errors at times. I think that it should be ActiveRecord's job to not try to `const_get` a class that isn't allowed.

```ruby
class SharedItem < ApplicationRecord
  belongs_to :item, polymorphic: true
  belongs_to :user

  validates :item_type, inclusion: { in: %w[Foo Bar Blah] }
end
```

```ruby
SharedItem.new(item_type: 'Fake', item_id: 1).valid?
(irb):17:in `<main>': uninitialized constant Fake (NameError)

      Object.const_get(camel_cased_word)
```

That occurs on a model with an inclusion validator on `item_type`. The `item` association is not optional, so Rails is attempting to see if the associated record exists. Seems like Rails should not be doing `Object.const_get()` on a string that wasn't in my validation's list. That's something I can add to this PR as well.

Our app makes heavy use of polymorphism, so I can think of like 6 models that I'd use this feature on. That's actually what got me to make this PR.

natematykiewicz commented:
Oh, the other part that I'm not sure if you caught in the conversation. This PR plays nicely with `optional: true`. If you've got `optional: true`, then the validation either needs to include a `nil` in the list of allowed values, or the validation needs an `if:` to only run if `item_type.present?`. This PR opted for the latter.

rafaelfranca commented:
I'm not sure how this is type checking, since this is just an inclusion validation, but I'm not sold in the syntax sugar yet, so I'm happy to reject the PR @dhh.

This feature is equivalent to add the following lines to a model:

```ruby
validates :item_type, inclusion: { in: %w(Post Comment) }, allow_blank: true
```

To properly build this syntax sugar, we would have to indeed change how the presence validation of associations work to take in consideration the same set of possible values. This is approach the level of complexity I'm not comfortable with, and I don't think it makes sense to add only the inclusion validation when an exception would be raised for any value outside of acceptable set.

dhh commented:
I see. Fair point with needing to validate the _type coming from a form to prevent a 500. So I'll retract my objection on the API side of thing. But I agree that the implementation needs to be proportionate to the problem. If we can find one that is, OK.

rafaelfranca commented:
@natematykiewicz can you add a new commit fixing the `Object.const_get()` problem to this PR? We can use that to evaluate how much complexity that part will add

natematykiewicz committed:
Prevent "NameError: unitialized constant" when provided bad user input

natematykiewicz commented:
I've added commit https://github.com/rails/rails/pull/51738/commits/6cbefac823e74128cf33c5e5d6ee8a76abb48712 to prevent the `NameError` when provided an unknown constant string.

Basically, it doesn't attempt to look up the `_type` constant unless `polymorphic` is `true` or the value of the `_type` field is in the array.

pdobb commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,31 @@
+
+    *Nate Matykiewicz*
+
+*   Added option to validate which models are allowed to be associated to a polymorphic `belongs_to`.
```
`Added` -> `Add` for present tense? I think this is the norm.

natematykiewicz replied:
Done

fxn commented:
Furthermore, also for consistency with the rest of associations including non-polymorphic `belongs_to`, the option I have in mind is `:class_names`. That would be round for my taste.

natematykiewicz commented:
@fxn so in my diff so far I was working towards:

```ruby
belongs_to :entryable, polymorphic: ["Message", "Comment"]
```

You're suggesting leaving the `polymorphic` option alone, and instead adding a `class_names` option. Like this, right?
```ruby
belongs_to :entryable, polymorphic: true, class_names: ["Message", "Comment"]
```

fxn commented:
@natematykiewicz yes, I am leaning on

```ruby
belongs_to :entryable, polymorphic: true, class_names: ["Message", "Comment"]
```

On one hand, it feels coherent with the rest of the association interfaces. You have the regular, short version, but you can add information if the macro cannot infer the class name/s.

Then, I am coming to this one from an introspection point of view. When working with reflections, I'd like the code to look like

```ruby
if relation.polymorphic?
  if klasses = relation.klasses
    # do something with the classes
  end
end
```

or something like that, without needing to do `polymorphic.is_a?(Array)`.

However, I might have a stab at it crediting you if we like the solution, for agility.

What makes me pause is the the behavior at the attribute level. If you receive the type in an API parameter, do you have validation error messages to present to the user? Should we raise in the associations writer API, but validate the low level attribute instead of raising also there?

natematykiewicz commented:
I like what you have better.

If you want to try implementing that, that'd be great. I'm a bit lacking in free time right now, so I'm not sure when I'd be able to revise this PR otherwise.

natematykiewicz commented:
As far as the interface goes, I like raising `ActiveRecord::AssociationTypeMismatch` on assignment because it consistent with other association types.

fxn commented:
@natematykiewicz cools. In your case, would work for a bare `foo.bar_type = params[:bar_type]` API-style assignment that does not go through the association API?

Because we have two levels of interfaces to think about, you know, `foo.bar = bar_record` and `foo.bar_type = "Bar"`. First one is clear, raise, I am not 100% sure about what would be the right thing to do for the second one.

fxn commented:
> What should klasses return for non-polymorphic associations? [klass] or nil? The former might make implementation considerably simpler.

Excellent question. My initial thought would be that other reflections would not respond to `klasses` (or would raise), so you have to guard the call, like you have to guard `klass` nowadays.

Returning `[klass]` would be perfect from a logical sense, thought about that too. But I am not sure how practical or confusing might that be.

Another option is that `klass` returns an array in polymorphics. This one is simple to understand, but the singular is weird.

I have not arrived to a conclusion here.
